framework/decode/custom_ags_ascii_consumer.cpp

```diff
@@ -0,0 +1,701 @@
+                              tab_size,
+                              StringOfDevices(gpuInfo.numDevices, returned_devices, tab_count, tab_size));
+            }));
+    });
+}
+
+void AgsAsciiConsumer::Process_agsDriverExtensionsDX12_CreateDevice(
+    const ApiCallInfo&                                    call_info,
+    AGSReturnCode                                         return_value,
+    AGSContext*                                           context,
+    AGSDX12DeviceCreationParams*                          creationParams,
+    const AGSDX12ExtensionParams*                         extensionParams,
+    StructPointerDecoder<Decoded_AGS_DX12_RETURN_PARAMS>* pReturnParams)
+{
+    using namespace gfxrecon::util;
+    uint32_t               tab_count = 0;
+    uint32_t               tab_size  = 4;
+    WriteAgsCallToFileInfo writeApiCallToFileInfo{};
+    writeApiCallToFileInfo.block_index   = current_block_index_;
+    writeApiCallToFileInfo.pFunctionName = "agsDriverExtensionsDX12_CreateDevice";
+    std::string returnValue              = util::ToString(return_value, to_string_flags_, tab_count, tab_size);
+    writeApiCallToFileInfo.pReturnValue  = !returnValue.empty() ? returnValue.c_str() : nullptr;
+    WriteApiCallToFile(writeApiCallToFileInfo, tab_count, tab_size, [&](std::stringstream& str_strm) {
+        FieldToString(str_strm, true, "ags context", to_string_flags_, tab_count, tab_size, PtrToString(context));
+
+        FieldToString(str_strm,
+                      false,
+                      "creationParams",
+                      to_string_flags_,
+                      tab_count,
+                      tab_size,
+                      ObjectToString(to_string_flags_, tab_count, tab_size, [&](std::stringstream& strStrm) {
+                          FieldToString(strStrm,
+                                        true,
+                                        "pAdapter",
+                                        to_string_flags_,
+                                        tab_count,
+                                        tab_size,
+                                        PtrToString(creationParams->pAdapter));
+                          FieldToString(strStrm,
+                                        false,
+                                        "riid",
+                                        to_string_flags_,
+                                        tab_count,
+                                        tab_size,
+                                        ToString(creationParams->iid, to_string_flags_, tab_count, tab_size));
+                          FieldToString(strStrm,
+                                        false,
+                                        "FeatureLevel",
+                                        to_string_flags_,
+                                        tab_count,
+                                        tab_size,
+                                        FeatureLevelToString(creationParams->FeatureLevel));
```

Comment:
Can this also use the `ToString` method from `generated_dx12_enum_to_string.h` and the function `std::string FeatureLevelToString(const D3D_FEATURE_LEVEL& value)` above be removed?

Reply:
Done. Sorry for the delay.